From 7e11e1e5439effc4af6e3a06314614a68ed3fb7f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michal=20=C4=8Ciha=C5=99?= <michal@cihar.com>
Date: Tue, 16 Dec 2025 16:20:43 +0100
Subject: [PATCH 1/1] fix(component): do not try to reacquire not held lock

The linked component lock does not have to be held in some contexts.

Fixes WEBLATE-38AZ
---
 weblate/trans/models/component.py | 2 +-
 weblate/utils/lock.py             | 1 +
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/weblate/trans/models/component.py b/weblate/trans/models/component.py
index 19ea646d7d..69c4b29fd5 100644
--- a/weblate/trans/models/component.py
+++ b/weblate/trans/models/component.py
@@ -2759,7 +2759,7 @@ class Component(
     def refresh_lock(self) -> None:
         """Refresh the lock to avoid expiry in long operations."""
         self.lock.reacquire()
-        if self.linked_component:
+        if self.linked_component and self.linked_component.lock.is_locked:
             self.linked_component.lock.reacquire()

     def _create_translations(  # noqa: C901,PLR0915
diff --git a/weblate/utils/lock.py b/weblate/utils/lock.py
index 4ebacdab53..f4779647d6 100644
--- a/weblate/utils/lock.py
+++ b/weblate/utils/lock.py
@@ -111,6 +111,7 @@ class WeblateLock:

         This is needed with Redis as the lock is expiring to avoid it stay infinitely.
         """
+        self.add_breadcrumb("reacquire")
         if self._using_redis:
             self._redis_lock.reacquire()

--
2.47.3
